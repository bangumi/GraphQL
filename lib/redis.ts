import IORedis from 'ioredis';

import { redisOption } from './config';

export default new IORedis(redisOption);
export const Subscriber = new IORedis(redisOption);
