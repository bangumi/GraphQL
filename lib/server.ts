import * as fs from 'node:fs';
import * as path from 'node:path';

import type { FastifyInstance, FastifyRequest, FastifyServerOptions } from 'fastify';
import { fastify } from 'fastify';
import mercurius from 'mercurius';
import AltairFastify from 'altair-fastify-plugin';
import swagger from '@fastify/swagger';
import type { OpenAPIV3 } from 'openapi-types';

import { schema } from './graphql/schema';
import type { Context } from './graphql/context';
import prisma from './prisma';
import * as auth from './auth';
import * as rest from './rest';
import { pkg, projectRoot } from './config';
import { Security } from './openapi';

export async function createServer(opts: FastifyServerOptions = {}): Promise<FastifyInstance> {
  const server = fastify(opts);

  await server.register(mercurius, {
    schema,
    path: '/v0/graphql',
    graphiql: false,
    allowBatchedQueries: true,
    context: async (request: FastifyRequest): Promise<Context> => {
      return {
        prisma,
        auth: await auth.byHeader(request.headers.authorization),
      };
    },
  });

  await server.register(AltairFastify, {
    path: '/v0/altair/',
    baseURL: '/v0/altair/',
    endpointURL: '/v0/graphql',
    initialSettings: {
      theme: 'dark',
      'plugin.list': ['altair-graphql-plugin-graphql-explorer'],
    },
  });

  const swaggerUI = fs.readFileSync(path.join(projectRoot, './lib/swagger.html'));

  server.get('/v0.5/', (_, res) => {
    void res.type('text/html').send(swaggerUI);
  });

  server.get('/v0.5/openapi.json', () => {
    return server.swagger();
  });

  const openapi: Partial<OpenAPIV3.Document> = {
    info: {
      version: pkg.version,
      title: 'hello',
    },
    components: {
      securitySchemes: {
        [Security.OptionalHTTPBearer]: {
          type: 'http',
          description:
            '不强制要求用户认证，但是可能看不到某些敏感内容内容（如 NSFW 或者仅用户自己可见的收藏）',
          scheme: 'Bearer',
        },
        [Security.HTTPBearer]: {
          type: 'http',
          description: '需要使用 access token 进行认证',
          scheme: 'Bearer',
        },
      },
    },
  };

  await server.register(swagger, {
    openapi,
  });

  await server.register(rest.setup, { prefix: '/v0.5' });

  return server;
}
