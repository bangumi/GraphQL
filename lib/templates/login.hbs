<script src='https://js.hcaptcha.com/1/api.js?hl=zh-CN' defer></script>
<div class='row'>
  <ul class='list-group error-details'></ul>
</div>
<div class='row'>
  <form>
    <div class='mb-3'>
      <label for='exampleInputEmail1' class='form-label'>Email address</label>
      <input
        type='email'
        class='form-control'
        name='email'
        id='exampleInputEmail1'
        aria-describedby='emailHelp'
      />
      <div id='emailHelp' class='form-text'>email</div>
    </div>
    <div class='mb-3'>
      <label for='exampleInputPassword1' class='form-label'>Password</label>
      <input
        type='password'
        name='password'
        class='form-control'
        id='exampleInputPassword1'
      />
    </div>
    <div class='mb-3'>
      {{#if production}}
        <div class='h-captcha' data-sitekey='4874acee-9c6e-4e47-99ad-e2ea1606961f'></div>
      {{else}}
        <div class='h-captcha' data-sitekey='10000000-ffff-ffff-ffff-000000000001'></div>
      {{/if}}
    </div>
    <button type='submit' class='btn btn-primary'>Submit</button>
  </form>
</div>
<script>
  $(document).ready(function() {
    $("form").on("submit", function(e) {
      e.preventDefault();
      $(".error-details").html("");
      const data = {};
      Array.from($("form").serializeArray()).forEach(({ name, value }, index) => {
        data[name] =
          value;
      });

      async function submit() {
        const res = await fetch('/p1/login', {
          method: "post",
          headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(data),
        });
        if
        (res.status > 300) {
          hcaptcha.reset();
          const d = await res.json();
          if (res.status === 401) {
            const remain = res.headers.get('X-RateLimit-Remaining');
            $(".error-details").append(`<li
      class="m-1 list-group-item list-group-item-danger">${d.message}</li>`).append(`<li class="m-1
      list-group-item list-group-item-danger">剩余登录次数: ${remain}</li>`);
          } else {
            console.log(d);
            d.details.forEach((detail, index) => {
              $(".error-details").prepend(`<li
      class="m-1 list-group-item list-group-item-danger">${detail}</li>`);
            });
          }
        } else {
          console.log(await res.json());
          location.href = '/p1/demo/';
        }
      }

      submit();
    });
  });
</script>
