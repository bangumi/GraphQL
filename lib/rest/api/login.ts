import * as crypto from 'node:crypto';

import * as bcrypt from 'bcrypt';
import type { Static } from '@sinclair/typebox';
import { Type as t } from '@sinclair/typebox';
import { createError } from '@fastify/error';
import httpCodes from 'http-status-codes';

import { redisPrefix } from '../../config';
import { Tag } from '../../openapi';
import redis from '../../redis';
import type { IUser } from '../../types/user';
import { User } from '../../types/user';
import prisma from '../../prisma';
import { randomBase62String } from '../../utils';
import type { App } from '../type';
import Limiter from '../../utils/rate-limit';

const CookieKey = 'sessionID';

const TooManyRequestsError = createError(
  'TOO_MANY_REQUESTS',
  'too many failed login attempts',
  httpCodes.TOO_MANY_REQUESTS,
);

const EmailPasswordNotMatch = createError(
  'EMAIL_PASSWORD_MISMATCH',
  'email does not exists or email and password not match',
  httpCodes.UNAUTHORIZED,
);

const emailPassMismatchError = t.Object({
  title: t.String(),
  description: t.String(),
  detail: t.Object({
    remain: t.Integer({ description: '剩余可用登录次数。' }),
  }),
});

export function setup(app: App) {
  // 10 calls per 600s
  const limiter = new Limiter({
    redisClient: redis,
    limit: 10,
    duration: 600,
  });

  app.post(
    '/login',
    {
      schema: {
        operationId: 'auth-login',
        tags: [Tag.Auth],
        response: {
          [httpCodes.OK]: t.Ref(User),
          [httpCodes.UNAUTHORIZED]: emailPassMismatchError,
        },
        body: t.Object({
          email: t.String({ minLength: 1 }),
          password: t.String({ minLength: 1 }),
        }),
      },
    },
    async function handler({ body: { email, password }, ip }, res): Promise<IUser> {
      const { remain, reset } = await limiter.get(`${redisPrefix}-login-rate-limit-${ip}`);
      if (remain <= 0) {
        void res.header('Retry-After', reset);
        throw new TooManyRequestsError();
      }

      const user = await prisma.members.findFirst({ where: { email } });

      if (!user) {
        return res.code(401).send({
          title: '',
          description: '',
          detail: { remain },
        } satisfies Static<typeof emailPassMismatchError>);
      }

      if (!(await comparePassword(user.password_crypt, password))) {
        throw new EmailPasswordNotMatch();
      }

      const now = Math.trunc(Date.now() / 1000);

      const token = randomBase62String(32);

      const value = {
        reg_time: user.regdate,
        user_id: user.id,
        created_at: now,
        expired_at: now + 60 * 60 * 24 * 7,
      };

      await prisma.chii_os_web_sessions.create({
        data: {
          value: Buffer.from(JSON.stringify(value)),
          user_id: user.id,
          created_at: value.created_at,
          expired_at: value.expired_at,
          key: token,
        },
      });

      void res.cookie(CookieKey, token, { sameSite: 'strict' });

      return { ID: user.id, username: user.username, nickname: user.nickname };
    },
  );
}

function processPassword(s: string): string {
  return crypto.createHash('md5').update(s).digest('hex');
}

export async function comparePassword(hashed: string, input: string): Promise<boolean> {
  return bcrypt.compare(processPassword(input), hashed);
}
