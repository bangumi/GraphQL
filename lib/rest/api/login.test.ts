import { beforeEach, describe, expect, test } from '@jest/globals';

import redis from '../../redis';
import { createServer } from '../../server';
import { comparePassword } from './login';

describe('login', () => {
  beforeEach(async () => {
    for (const key of await redis.keys('*')) {
      if (key.includes('1122')) {
        await redis.del(key);
      }
    }
  });

  test('should failed on too many requests', async () => {
    const app = await createServer();

    const opt = {
      method: 'post',
      url: '/v0.5/login',
      payload: { email: 'ee', password: 'eepp', 'h-captcha-response': 'fake-response' },
      headers: {
        'cf-connecting-ip': '1122',
      },
    } as const;

    const login = () => app.inject(opt);

    const all = await Promise.all(Array.from({ length: 20 }).map(() => login()));

    const res = await login();

    expect(all.filter((x) => x.statusCode === 429)).not.toHaveLength(0);

    expect(res.statusCode).toBe(429);
    expect(res.json()).toMatchSnapshot();
  });
});

describe('compare password', () => {
  test('should pass', async () => {
    const hashed = '$2a$12$GA5Pr9GhsyLJcSPoTpYBY.JqTzYZb2nfgSeZ1EK38bfgk/Rykkvuq';
    const input = 'lovemeplease';
    await expect(comparePassword(hashed, input)).resolves.toBe(true);
  });

  test('should not pass', async () => {
    const hashed = '$2a$12$GA5Pr9GhsyLJcSPoTpYBY.JqTzYZb2nfgSeZ1EK38bfgk/Rykkvuq';
    const input = 'lovemeplease1';
    await expect(comparePassword(hashed, input)).resolves.toBe(false);
  });
});
