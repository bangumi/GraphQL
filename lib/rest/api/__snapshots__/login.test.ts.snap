// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login should failed on too many requests 1`] = `
{
  "code": "TOO_MANY_REQUESTS",
  "error": "Too Many Requests",
  "message": "too many failed login attempts",
  "statusCode": 429,
}
`;
